internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+// Returns the collected results in a JobOutput.
+// If the context is canceled during execution, returns nil.
+// When not waiting for completion, the JobOutput may still be modified until the Done channel is closed.
+// Returning a copy of results gives immediate access to stable results but the final results will not be accessible later.
+func (this *Job[InputType, OutputType]) Run() *JobOutput[InputType, OutputType] {
+	done := make(chan any)
+
+	output := JobOutput[InputType, OutputType]{
+		Done:           done,
+		ResultItems:    make([]OutputType, 0, len(this.WorkItems)),
+		RemainingItems: this.WorkItems,
+		RunTime:        0,
+		WorkErrors:     make(map[int]error, 0),
+		JobID:          util.UUID(),
+	}
+
+	err := this.Validate()
+	if err != nil {
+		close(done)
+
+		output.Error = fmt.Errorf("Failed to validate job: '%v'.", err)
+		return &output
+	}
+
+	// If we are overwriting records, remove all the old records.
+	if this.OverwriteRecords && !this.DryRun && this.RemoveStorageFunc != nil {
+		output.Error = this.RemoveStorageFunc(this.WorkItems)
+		if output.Error != nil {
+			close(done)
+
+			return &output
+		}
+	}
+
+	// Query for stored records.
+	if !this.OverwriteRecords && this.RetrieveFunc != nil {
+		output.ResultItems, output.RemainingItems, output.Error = this.RetrieveFunc(this.WorkItems)
+		if output.Error != nil {
+			close(done)
+
+			return &output
+		}
+	}
+
+	if this.WaitForCompletion {
+		this.run(&output, true)
+
+		close(done)
+
+		// If the context was canceled during execution, return immediately.
+		if this.Context.Err() != nil {
+			return nil
+		}
+	} else if this.ReturnCopy {
+		go func() {
+			backgroundOutput := &JobOutput[InputType, OutputType]{
+				RemainingItems: output.GetRemainingItems(),
+			}
+
+			this.run(backgroundOutput, false)
+
+			if backgroundOutput.Error != nil {
+				log.Error("Failure while running asynchronous job: '%v'.", output.Error)
+			}
+		}()
+
+		close(done)
+		return &output
+	} else {
+		go func() {
+			this.run(&output, true)
+			if output.Error != nil {
+				log.Error("Failure while running asynchronous job: '%v'.", output.Error)
+			}
+
+			close(done)
+		}()
+	}
+
+	return &output
+}
+
+// Given a partial job output, run job processes the remaining items in a thread safe manner.
+// When not updating the output, only the run time and errors will be available.
+func (this *Job[InputType, OutputType]) run(output *JobOutput[InputType, OutputType], updateOutput bool) {
+	if len(output.GetRemainingItems()) == 0 {
+		return
+	}
+
+	noLockWait := true
+	if this.LockKey != "" {
+		noLockWait = lockmanager.Lock(this.LockKey)
+		defer lockmanager.Unlock(this.LockKey)
+	}
+
+	// The context has been canceled while waiting for a lock, abandon this job.
+	if this.Context.Err() != nil {
+		return
+	}
+
+	var err error = nil
+
+	// If we had to wait for the lock, then check again for stored records.
+	if !noLockWait && !this.OverwriteRecords && this.RetrieveFunc != nil {
```

Comment:
I don't think it is necessary to check for `OverwriteRecords` here.
I think the only check you need is to delete records in the very beginning
(and then you have it on the very first retrieval, but that is just for performance).

The records got dropped, and therefore they were overwritten.
If another job happened be the one ti place the records after they were deleted,
that is just fine.